### ISSUE TITLE
[New Edit Mode] Add Elements pool to New Edit Mode

### ISSUE THREAD

**Issue by charlesh88** [nasa/openmctweb#279]
From https://github.com/nasa/openmctweb/issues/161#issuecomment-155162928 No. 3

Bring across elements pool in Inspector during Edit mode - user can invoke context actions on items, principally to remove elements.

UI sketches have been updated at https://drive.google.com/a/nasa.gov/file/d/0B0v27OM1JkSeR001RFJZMk43QVk/view?usp=sharing. Note that the design utilizes a split pane approach within the inspector.
Charles will initially stub in the markup for this, then reassign for wiring up.

**Comment by akhenry** [comment]
Comments (from https://github.com/nasa/openmctweb/issues/161 )

>User is editing a layout that contains the original of a panel. The panel has multiple links of itself elsewhere, in folders and other objects. The user deletes the original using the Elements pool of the layout. What happens in the tree?
@charlesh88 

>Technically, we don't have delete yet, so I'd say we punt on this. But, if we had delete and didn't consider this a edge case, deleting an original in a layout would break it.
@larkin

>I think the Delete should not take effect until the user Saves. It would be really scary to me as a user if some of my actions in Edit mode were immediate, and others could be cancelled.
>If the user drags the object back in after a Delete, though, well, that's weird. Do we say "that thing doesn't exist in the little version of the world that you're editing"? Do we cancel the Delete?
>After Save, I think all links in the tree should be replaced with Missing Objects.
@VWoeltjen 

>Totally agree on deletion not taking effect until a Save event on the object being edited.
>On the question about what happens in the tree when an original is deleted from another object being edited, I'm fine with the approach that the tree doesn't update until a Save event occurs in the object being edited.
>That said, the scenario where a user drags a deleted original back in to an object being edited is weird. Not sure what to do here, and I think that we'll need to dig more into this when deletion is being implemented.
>As for what happens to the links that point at a deleted original, when we begin to implement deletion, we should re-examine how we want to handle this: there are pros and cons to both approaches of replacing links with missing objects (lots of leftover cruft) or just deleting the links too (WTF happened to my links?).
@charlesh88

**Comment by akhenry** [comment]
## On dragging an object or its child onto itself
(from #161)
>Dragging an object into itself should be disallowed.
>Can the user drag an object's child into itself again? I think you should, the action would add another linked version of the object being dragged (once/if we allow multiple instances of the same object in a layout, for example).
@charlesh88 

>Hm. Multiple linked instances of the same object seems weird to me, and we'd have to change the way we model location to accommodate this. I do think we should allow multiple occurrences of the same object in frames in a Display Layout, but I don't see the need to represent them multiple times in the Elements pool or the tree.
@VWoeltjen 

>Multiple instances of a single object in a layout was the use case that was top of mind for me. I do think they would need to be represented multiple times in the Elements pool, but not in the tree. The rationale here is the role that an enhanced Inspector could play in edit mode: all objects (including boxes and lines) might manifest in the Elements pool. The notion is that the Elements pool is just an alternate representation of everything within an object.
@charlesh88

**Comment by charlesh88** [comment]
Bring in updated open master and convert l-flex to new flexbox strategy.

**Comment by charlesh88** [comment]
open279 should now be deleted; this branch has been integrated into open199-meet-open279 and work on this issue should take place there moving forward.

**Comment by akhenry** [comment]
Have implemented this and pushed to branch open279-proto. Have also merge all recent changes into open199 so they can be previewed by others.

The context menu on objects in the pool has all of the usual options from the navigation tree. All context menu options work, although I do observe some strange behaviour with some of them (detailed further down). I'd like to check with you guys whether we want to have all of these context options enabled initially -
<img width="140" alt="screen shot 2015-11-25 at 11 32 00 am" src="https://cloud.githubusercontent.com/assets/4215777/11407535/49fbd638-9368-11e5-851c-86c4ed5e76b5.png">

Some notes:

1. Object pool does not show hierarchy yet. I think this is ok for now, as we don't have in-place editing of child objects.
1. The 'move' menu option creates a link. I'm not sure why this is. If we want to retain this functionality, I will do some investigation
1. If in a layout I edit the properties of a clock and change the date format, it is not updated until I save the enclosing layout. Other properties that I tested, such as name are updated immediately.
1. The reimplemented duplication is not really compatible with edit mode, although outwardly it appears to work. If I duplicate an object in a layout it appears in the layout, and if I save it is persisted. If I cancel instead of saving, the duplicated object does not appear in the layout. Behind the scenes though, duplicate will always create and persist objects to the store, so this will create orphans if edit mode is cancelled. Duplication needs refactoring to use object capabilities so that they can use the caching capabilities of the EditableDomainObject.
1. @charlesh88 I had to mode the search text field into the elements.html view. This has resulted in reduced spacing between the text field and the first element in the pool, and probably needs some css tweaking.

**Comment by charlesh88** [comment]
This also maps to #138, which asks for context menu "contexts".

**Comment by akhenry** [comment]
We need to decide what subset of context menu items we want to enable for NEM beta. I think the minimum we can get away with is remove, edit properties would be nice though as well.

**Comment by larkin** [comment]
@akhenry the current edit mode does not enforce a subset on the menu so I'd consider this not a blocker for NEM beta.

Current:
![screen shot 2015-12-08 at 2 31 05 pm](https://cloud.githubusercontent.com/assets/226503/11670733/66eade2e-9db8-11e5-8940-3ae1a94bb69e.png)

New:
![screen shot 2015-12-08 at 2 31 17 pm](https://cloud.githubusercontent.com/assets/226503/11670736/69d9ac14-9db8-11e5-99c4-81e5a1909d13.png)

**Comment by akhenry** [comment]
Ok, interesting. We will need to investigate 2,3,4,5 from https://github.com/nasa/openmctweb/issues/279#issuecomment-159710699 to establish which if any of these are blocking issues and how we want to deal with them.

**Comment by larkin** [comment]
I imagine 2,3,4 and 5 are not new behaviors-- they probably have the same effects in the current edit mode, so I wouldn't consider them blockers.  I think we should move this discussion to a new issue and close this issue as it has been completed.

**Comment by charlesh88** [comment]
Agree with Pete as long as move this to a new issue (NEM Elements pool refinements, or some such) or file separate issues for 2,3,4 and 5 as above. I'd close this only after this is done so we don't lose sight of those issues.

**Comment by akhenry** [comment]
I have created tickets for the outstanding issues above. Created them as separate issues to make it easier to track them, but also because I think at least one of them might need to be fixed for NEM Beta, but the rest probably don't as they either don't have use impact, or they are existing issues anyway.

#396 is the one issue I think probably should be fixed for NEM Beta
